File: runtime/src/main/java/com/siimkinks/sqlitemagic/Update.java
package com.siimkinks.sqlitemagic;

import android.support.annotation.*;

import com.siimkinks.sqlitemagic.Select.Select1;

import java.util.ArrayList;
import java.util.Collections;

/**
 * Builder for SQL UPDATE statement.
 */
public final class Update extends UpdateSqlNode {
  Update() {
    super(null);
  }

  @Override
  protected void appendSql(@NonNull StringBuilder sb) {
    sb.append("UPDATE");
  }

  /**
   * Create a new builder for SQL UPDATE statement with conflict algorithm.
   *
   * @param conflictAlgorithm Conflict resolution algorithm to use
   * @return A new builder for SQL UPDATE statement
   */
  @CheckResult
  public static UpdateConflictAlgorithm withConflictAlgorithm(@ConflictAlgorithm int conflictAlgorithm) {
    return new UpdateConflictAlgorithm(new Update(), conflictAlgorithm);
  }

  /**
   * Create a new builder for SQL UPDATE statement.
   * <p>
   * Note that {@code table} param must be one of annotation processor generated table
   * objects that corresponds to table in a database.<br>
   * Example:
   * <pre>{@code
   * import static com.example.model.AuthorTable.AUTHOR;
   *
   * // [...]
   *
   * Update.table(AUTHOR)...
   * }</pre>
   *
   * @param table Table to update. This param must be one of annotation processor
   *              generated table objects that corresponds to table in a database
   * @param <T>   Table object type
   * @return A new builder for SQL UPDATE statement
   */
  @CheckResult
  public static <T> TableNode<T> table(@NonNull Table<T> table) {
    return new TableNode<>(new Update(), table.name);
  }

  /**
   * Create a new builder for SQL UPDATE statement.
   * <p>
   * Example:
   * <pre>{@code
   * Update.table("author")...
   * }</pre>
   *
   * @param tableName Table to update
   * @return A new builder for SQL UPDATE statement
   */
  @CheckResult
  public static TableNode<?> table(@NonNull String tableName) {
    return new TableNode<>(new Update(), tableName);
  }

  /**
   * Builder for SQL UPDATE statement.
   */
  public static final class UpdateConflictAlgorithm extends UpdateSqlNode {
    @ConflictAlgorithm
    private final int conflictAlgorithm;

    UpdateConflictAlgorithm(@NonNull Update parent, @ConflictAlgorithm int conflictAlgorithm) {
      super(parent);
      this.conflictAlgorithm = conflictAlgorithm;
    }

    @Override
    protected void appendSql(@NonNull StringBuilder sb) {
      sb.append(ConflictAlgorithm.CONFLICT_VALUES[conflictAlgorithm]);
    }

    /**
     * Define a table to update.
     * <p>
     * Note that {@code table} param must be one of annotation processor generated table
     * objects that corresponds to table in a database.<br>
     * Example:
     * <pre>{@code
     * import static com.example.model.AuthorTable.AUTHOR;
     *
     * // [...]
     *
     * Update.table(AUTHOR)...
     * }</pre>
     *
     * @param table Table to update. This param must be one of annotation processor
     *              generated table objects that corresponds to table in a database
     * @param <T>   Table object type
     * @return SQL UPDATE statement builder
     */
    @CheckResult
    public <T> TableNode<T> table(@NonNull Table<T> table) {
      return new TableNode<>(this, table.name);
    }

    /**
     * Define a table to update.
     * <p>
     * Example:
     * <pre>{@code
     * Update.table("author")...
     * }</pre>
     *
     * @param tableName Table to update
     * @return SQL UPDATE statement builder
     */
    @CheckResult
    public TableNode<?> table(@NonNull String tableName) {
      return new TableNode<>(this, tableName);
    }
  }

  /**
   * Builder for SQL UPDATE statement.
   *
   * @param <T> Updated table object type
   */
  public static final class TableNode<T> extends UpdateSqlNode {
    @NonNull
    final String tableName;

    TableNode(@NonNull UpdateSqlNode parent, @NonNull String tableName) {
      super(parent);
      this.tableName = tableName;
      updateBuilder.tableNode = this;
    }

    @Override
    protected void appendSql(@NonNull StringBuilder sb) {
      sb.append(tableName);
    }

    /**
     * Update a not nullable column with new not nullable value.
     *
     * @param column Column to update. This param must be one of annotation processor
     *               generated column objects that corresponds to column in a database
     *               table
     * @param value  A new value to set for updated column
     * @param <V>    Value type
     * @param <R>    Column return type
     * @param <ET>   Column equivalent type
     * @return SQL UPDATE statement builder
     */
    @CheckResult
    public <V, R, ET> Set<T> set(@NonNull Column<V, R, ET, T, NotNullable> column, @NonNull V value) {
      return new Set<>(this, new UpdateColumn<>(column).is(value));
    }

    /**
     * Update a nullable column with new nullable value.
     *
     * @param column Column to update. This param must be one of annotation processor
     *               generated column objects that corresponds to column in a database
     *               table
     * @param value  A new value to set for updated column
     * @param <V>    Value type
     * @param <R>    Column return type
     * @param <ET>   Column equivalent type
     * @return SQL UPDATE statement builder
     */
    @CheckResult
    public <V, R, ET> Set<T> setNullable(@NonNull Column<V, R, ET, T, Nullable> column, @android.support.annotation.Nullable V value) {
      return new Set<>(this, new UpdateColumn<>(column).isNullable(value));
    }

    /**
     * Update a complex column with new value.
     *
     * @param column Column to update. This param must be one of annotation processor
     *               generated complex column objects that corresponds to a complex column
     *               in a database table
     * @param value  A new value to set for updated column
     * @param <V>    Value type
     * @param <R>    Column return type
     * @param <ET>   Column equivalent type
     * @param <N>    Column nullability
     * @return SQL UPDATE statement builder
     */
    @CheckResult
    public <V, R, ET, N> Set<T> set(@NonNull ComplexColumn<V, R, ET, T, N> column, long value) {
      return new Set<>(this, new UpdateColumn<>(column).is(value));
    }

    /**
     * Update a column with the value of another column.
     *
     * @param column           Column to update. This param must be one of annotation processor
     *                         generated column objects that corresponds to column in a database
     *                         table
     * @param assignmentColumn Column who's value to assign to {@code column}.
     *                         This param must be one of annotation processor
     *                         generated column objects that corresponds to column in a database
     *                         table
     * @param <V>              Column value type
     * @param <R>              Column return type
     * @param <ET>             Column equivalent type
     * @param <N>              Column nullability
     * @return SQL UPDATE statement builder
     */
    @CheckResult
    public <V, R, ET, N> Set<T> set(@NonNull Column<V, R, ET, T, N> column,
                                    @NonNull Column<?, ?, ? extends ET, ?, ? super N> assignmentColumn) {
      return new Set<>(this, new UpdateColumn<>(column).is(assignmentColumn));
    }

    /**
     * Update a column with the value of inner SELECT statement result.
     *
     * @param column Column to update. This param must be one of annotation processor
     *               generated column objects that corresponds to column in a database
     *               table
     * @param select Inner select statement who's result will be assigned to {@code column}.
     *               Select statement must return 1x1 result set.
     * @param <V>    Column value type
     * @param <R>    Column return type
     * @param <ET>   Column equivalent type
     * @param <N>    Column nullability
     * @return SQL UPDATE statement builder
     */
    @CheckResult
    public <V, R, ET, N> Set<T> set(@NonNull Column<V, R, ET, T, N> column,
                                    @NonNull SelectSqlNode.SelectNode<? extends ET, Select1, ? super N> select) {
      return new Set<>(this, new UpdateColumn<>(column).is(select));
    }

    /**
     * Update a column with new value.
     *
     * @param columnName Column to update
     * @param value  A new value to set for updated column
     * @return SQL UPDATE statement builder
     */
    @CheckResult
    public Set<T> set(@NonNull String columnName, @android.support.annotation.Nullable String value) {
      return new Set<>(this, columnName, value);
    }
  }

  private static final class UpdateColumn<T, R, ET, P, N> extends ComplexColumn<T, R, ET, P, N> {
    @NonNull
    private final Column<T, R, ET, P, N> parentColumn;

    UpdateColumn(@NonNull Column<T, R, ET, P, N> parentColumn) {
      super(parentColumn.table, parentColumn.name, parentColumn.allFromTable, parentColumn.valueParser, parentColumn.nullable, parentColumn.alias);
      this.parentColumn = parentColumn;
    }

    Expr isNullable(@android.support.annotation.Nullable T value) {
      final String nullableValue = value != null ? toSqlArg(value) : null;
      return new Expr1(this, "=?", nullableValue);
    }

    @NonNull
    @Override
    String toSqlArg(@NonNull T val) {
      return parentColumn.toSqlArg(val);
    }

    @Override
    void appendSql(@NonNull StringBuilder sb) {
      sb.append(name);
    }
  }

  /**
   * Builder for SQL UPDATE statement.
   *
   * @param <T> Updated table object type
   */
  public static final class Set<T> extends UpdateNode {
    private final ArrayList<Expr> updates = new ArrayList<>(1);
    private final ArrayList<String> rawUpdates = new ArrayList<>();

    Set(@NonNull UpdateSqlNode parent, @NonNull Expr firstUpdate) {
      super(parent);
      updates.add(firstUpdate);
      firstUpdate.addArgs(updateBuilder.args);
    }

    Set(@NonNull UpdateSqlNode parent, @NonNull String columnName, @android.support.annotation.Nullable String value) {
      super(parent);
      rawUpdates.add(columnName);
      updateBuilder.args.add(value);
    }

    @Override
    protected void appendSql(@NonNull StringBuilder sb) {
      sb.append("SET ");
      final ArrayList<Expr> updates = this.updates;
      int size = updates.size();
      for (int i = 0; i < size; i++) {
        if (i != 0) {
          sb.append(',');
        }
        updates.get(i).appendToSql(sb);
      }
      final ArrayList<String> rawUpdates = this.rawUpdates;
      size = rawUpdates.size();
      for (int i = 0; i < size; i++) {
        if (i != 0) {
          sb.append(',');
        }
        sb.append(rawUpdates.get(i));
        sb.append("=?");
      }
    }

    /**
     * Update a not nullable column with new not nullable value.
     *
     * @param column Column to update. This param must be one of annotation processor
     *               generated column objects that corresponds to column in a database
     *               table
     * @param value  A new value to set for updated column
     * @param <V>    Value type
     * @param <R>    Column return type
     * @param <ET>   Column equivalent type
     * @return SQL UPDATE statement builder
     */
    @CheckResult
    public <V, R, ET> Set<T> set(@NonNull Column<V, R, ET, T, NotNullable> column, @NonNull V value) {
      final Expr expr = new UpdateColumn<>(column).is(value);
      updates.add(expr);
      expr.addArgs(updateBuilder.args);
      return this;
    }

    /**
     * Update a nullable column with new nullable value.
     *
     * @param column Column to update. This param must be one of annotation processor
     *               generated column objects that corresponds to column in a database
     *               table
     * @param value  A new value to set for updated column
     * @param <V>    Value type
     * @param <R>    Column return type
     * @param <ET>   Column equivalent type
     * @return SQL UPDATE statement builder
     */
    @CheckResult
    public <V, R, ET> Set<T> setNullable(@NonNull Column<V, R, ET, T, Nullable> column, @android.support.annotation.Nullable V value) {
      final Expr expr = new UpdateColumn<>(column).isNullable(value);
      updates.add(expr);
      expr.addArgs(updateBuilder.args);
      return this;
    }

    /**
     * Update a complex column with new value.
     *
     * @param column Column to update. This param must be one of annotation processor
     *               generated complex column objects that corresponds to a complex column
     *               in a database table
     * @param value  A new value to set for updated column
     * @param <V>    Value type
     * @param <R>    Column return type
     * @param <ET>   Column equivalent type
     * @param <N>    Column nullability
     * @return SQL UPDATE statement builder
     */
    @CheckResult
    public <V, R, ET, N> Set<T> set(@NonNull ComplexColumn<V, R, ET, T, N> column, long value) {
      final Expr expr = new UpdateColumn<>(column).is(value);
      updates.add(expr);
      expr.addArgs(updateBuilder.args);
      return this;
    }

    /**
     * Update a column with the value of another column.
     *
     * @param column           Column to update. This param must be one of annotation processor
     *                         generated column objects that corresponds to column in a database
     *                         table
     * @param assignmentColumn Column who's value to assign to {@code column}.
     *                         This param must be one of annotation processor
     *                         generated column objects that corresponds to column in a database
     *                         table
     * @param <V>              Column value type
     * @param <R>              Column return type
     * @param <ET>             Column equivalent type
     * @param <N>              Column nullability
     * @return SQL UPDATE statement builder
     */
    @CheckResult
    public <V, R, ET, N> Set<T> set(@NonNull Column<V, R, ET, T, N> column,
                                    @NonNull Column<?, ?, ? extends ET, ?, ? super N> assignmentColumn) {
      final Expr expr = new UpdateColumn<>(column).is(assignmentColumn);
      updates.add(expr);
      expr.addArgs(updateBuilder.args);
      return this;
    }

    /**
     * Update a column with the value of inner SELECT statement result.
     *
     * @param column Column to update. This param must be one of annotation processor
     *               generated column objects that corresponds to column in a database
     *               table
     * @param select Inner select statement who's result will be assigned to {@code column}.
     *               Select statement must return 1x1 result set.
     * @param <V>    Column value type
     * @param <R>    Column return type
     * @param <ET>   Column equivalent type
     * @param <N>    Column nullability
     * @return SQL UPDATE statement builder
     */
    @CheckResult
    public <V, R, ET, N> Set<T> set(@NonNull Column<V, R, ET, T, N> column,
                                    @NonNull SelectSqlNode.SelectNode<? extends ET, Select1, ? super N> select) {
      final Expr expr = new UpdateColumn<>(column).is(select);
      updates.add(expr);
      expr.addArgs(updateBuilder.args);
      return this;
    }

    /**
     * Update a column with new value.
     *
     * @param columnName Column to update
     * @param value  A new value to set for updated column
     * @return SQL UPDATE statement builder
     */
    @CheckResult
    public Set<T> set(@NonNull String columnName, @android.support.annotation.Nullable String value) {
      rawUpdates.add(columnName);
      updateBuilder.args.add(value);
      return this;
    }

    /**
     * Define SQL UPDATE statement WHERE clause.
     *
     * @param expr WHERE clause expression
     * @return SQL UPDATE statement builder
     */
    @CheckResult
    public Where where(@NonNull Expr expr) {
      return new Where(this, expr);
    }

    /**
     * Define SQL UPDATE statement WHERE clause.
     *
     * @param whereClause WHERE clause
     * @param whereArgs   WHERE clause arguments
     * @return SQL UPDATE statement builder
     */
    @CheckResult
    public RawWhere where(@NonNull String whereClause, @android.support.annotation.Nullable String... whereArgs) {
      return new RawWhere(this, whereClause, whereArgs);
    }
  }

  /**
   * Builder for SQL UPDATE statement.
   */
  public static final class Where extends UpdateNode {
    @NonNull
    private final Expr expr;

    Where(@NonNull UpdateSqlNode parent, @NonNull Expr expr) {
      super(parent);
      this.expr = expr;
      expr.addArgs(updateBuilder.args);
    }

    @Override
    protected void appendSql(@NonNull StringBuilder sb) {
      sb.append("WHERE ");
      expr.appendToSql(sb);
    }
  }

  /**
   * Builder for SQL UPDATE statement.
   */
  public static final class RawWhere extends UpdateNode {
    @NonNull
    private final String clause;

    RawWhere(@NonNull UpdateSqlNode parent, @NonNull String clause, @android.support.annotation.Nullable String[] args) {
      super(parent);
      this.clause = clause;
      if (args != null && args.length > 0) {
        Collections.addAll(updateBuilder.args, args);
      }
    }

    @Override
    protected void appendSql(@NonNull StringBuilder sb) {
      sb.append("WHERE ");
      sb.append(clause);
    }
  }
}

File: runtime/src/main/kotlin/com/siimkinks/sqlitemagic/Update.kt
package com.siimkinks.sqlitemagic

import androidx.annotation.CheckResult
import com.siimkinks.sqlitemagic.Select.Select1
import com.siimkinks.sqlitemagic.SelectSqlNode.SelectNode

/** Builder for SQL UPDATE statement. */
class Update internal constructor() : UpdateSqlNode(null) {
  override fun appendSql(sb: StringBuilder) {
    sb.append("UPDATE")
  }

  companion object {
    /**
     * Create a new builder for SQL UPDATE statement with conflict algorithm.
     *
     * @param conflictAlgorithm Conflict resolution algorithm to use
     * @return A new builder for SQL UPDATE statement
     */
    @JvmStatic
    @CheckResult
    fun withConflictAlgorithm(
      @ConflictAlgorithm conflictAlgorithm: Int
    ) = UpdateConflictAlgorithm(
      parent = Update(),
      conflictAlgorithm = conflictAlgorithm
    )

    /**
     * Create a new builder for SQL UPDATE statement.
     *
     * Note that `table` param must be one of annotation processor generated table
     * objects that corresponds to table in a database.
     *
     * Example:
     * ```
     * import static com.example.model.AuthorTable.AUTHOR;
     *
     * // [...]
     *
     * Update.table(AUTHOR)...
     * ```
     *
     * @param table Table to update. This param must be one of annotation processor
     * generated table objects that corresponds to table in a database
     * @param T Table object type
     * @return A new builder for SQL UPDATE statement
     */
    @JvmStatic
    @CheckResult
    fun <T> table(table: Table<T>) = TableNode<T>(
      parent = Update(),
      tableName = table.name
    )

    /**
     * Create a new builder for SQL UPDATE statement.
     *
     * Example:
     * ```
     * Update.table("author")...
     * ```
     *
     * @param tableName Table to update
     * @return A new builder for SQL UPDATE statement
     */
    @JvmStatic
    @CheckResult
    fun table(tableName: String): TableNode<*> = TableNode<Any>(
      parent = Update(),
      tableName = tableName
    )
  }

  /** Builder for SQL UPDATE statement. */
  class UpdateConflictAlgorithm internal constructor(
    parent: Update,
    @ConflictAlgorithm
    private val conflictAlgorithm: Int
  ) : UpdateSqlNode(parent) {
    override fun appendSql(sb: StringBuilder) {
      sb.append(ConflictAlgorithm.CONFLICT_VALUES[conflictAlgorithm])
    }

    /**
     * Define a table to update.
     *
     * Note that `table` param must be one of annotation processor generated table
     * objects that corresponds to table in a database.
     *
     * Example:
     * ```
     * import static com.example.model.AuthorTable.AUTHOR;
     *
     * // [...]
     *
     * Update.table(AUTHOR)...
     * ```
     *
     * @param table Table to update. This param must be one of annotation processor
     * generated table objects that corresponds to table in a database
     * @param T Table object type
     * @return SQL UPDATE statement builder
     */
    @CheckResult
    fun <T> table(table: Table<T>) = TableNode<T>(
      parent = this,
      tableName = table.name
    )

    /**
     * Define a table to update.
     *
     * Example:
     * ```
     * Update.table("author")...
     * ```
     *
     * @param tableName Table to update
     * @return SQL UPDATE statement builder
     */
    @CheckResult
    fun table(tableName: String): TableNode<*> = TableNode<Any>(
      parent = this,
      tableName = tableName
    )
  }

  /**
   * Builder for SQL UPDATE statement.
   *
   * @param T Updated table object type
   */
  class TableNode<T> internal constructor(
    parent: UpdateSqlNode,
    @JvmField
    internal val tableName: String
  ) : UpdateSqlNode(parent) {
    init {
      updateBuilder.tableNode = this
    }

    override fun appendSql(sb: StringBuilder) {
      sb.append(tableName)
    }

    /**
     * Update a not nullable column with new not nullable value.
     *
     * @param column Column to update. This param must be one of annotation processor
     * generated column objects that corresponds to column in a database table
     * @param value A new value to set for updated column
     * @param V Value type
     * @param R Column return type
     * @param ET Column equivalent type
     * @return SQL UPDATE statement builder
     */
    @CheckResult
    fun <V, R, ET> set(
      column: Column<V, R, ET, T, NotNullable>,
      value: V & Any
    ): Set<T> = Set(
      parent = this,
      firstUpdate = UpdateColumn(column) IS value
    )

    /**
     * Update a nullable column with new nullable value.
     *
     * @param column Column to update. This param must be one of annotation processor
     * generated column objects that corresponds to column in a database table
     * @param value A new value to set for updated column
     * @param V Value type
     * @param R Column return type
     * @param ET Column equivalent type
     * @return SQL UPDATE statement builder
     */
    @CheckResult
    fun <V, R, ET> setNullable(
      column: Column<V, R, ET, T, Nullable>,
      value: V?
    ): Set<T> = Set(
      parent = this,
      firstUpdate = UpdateColumn(column).isNullable(value)
    )

    /**
     * Update a column with the value of another column.
     *
     * @param column Column to update. This param must be one of annotation processor
     * generated column objects that corresponds to column in a database table
     * @param assignmentColumn Column who's value to assign to `column`.
     * This param must be one of annotation processor generated column objects that corresponds to column in a database
     * table
     * @param V Column value type
     * @param R Column return type
     * @param ET Column equivalent type
     * @param N Column nullability
     * @return SQL UPDATE statement builder
     */
    @CheckResult
    fun <V, R, ET, N> set(
      column: Column<V, R, ET, T, N>,
      assignmentColumn: Column<*, *, out ET, *, in N>
    ): Set<T> = Set(
      parent = this,
      firstUpdate = UpdateColumn(column) IS assignmentColumn
    )

    /**
     * Update a column with the value of inner SELECT statement result.
     *
     * @param column Column to update. This param must be one of annotation processor
     * generated column objects that corresponds to column in a database table
     * @param select Inner select statement who's result will be assigned to `column`.
     * Select statement must return 1x1 result set.
     * @param V Column value type
     * @param R Column return type
     * @param ET Column equivalent type
     * @param N Column nullability
     * @return SQL UPDATE statement builder
     */
    @CheckResult
    fun <V, R, ET, N> set(
      column: Column<V, R, ET, T, N>,
      select: SelectNode<out ET, Select1, in N>
    ): Set<T> = Set(
      parent = this,
      firstUpdate = UpdateColumn(column) IS select
    )

    /**
     * Update a column with new value.
     *
     * @param columnName Column to update
     * @param value A new value to set for updated column
     * @return SQL UPDATE statement builder
     */
    @CheckResult
    fun set(
      columnName: String,
      value: String?
    ): Set<T> = Set(
      parent = this,
      columnName = columnName,
      value = value
    )
  }

  private class UpdateColumn<T, R, ET, P, N>(
    private val parentColumn: Column<T, R, ET, P, N>
  ) : Column<T, R, ET, P, N>(
    parentColumn.table,
    parentColumn.name,
    parentColumn.allFromTable,
    parentColumn.valueParser,
    parentColumn.nullable,
    parentColumn.alias
  ) {
    fun isNullable(value: T?): Expr = Expr1(this, "=?", value?.let(::toSqlArg))

    override fun toSqlArg(value: T & Any): String = parentColumn.toSqlArg(value)

    override fun appendSql(sb: StringBuilder) {
      sb.append(name)
    }
  }

  /**
   * Builder for SQL UPDATE statement.
   *
   * @param T Updated table object type
   */
  class Set<T> internal constructor(parent: UpdateSqlNode) : UpdateNode(parent) {
    private sealed interface Assignment {
      fun appendSql(sb: StringBuilder)

      class Expression(private val expr: Expr) : Assignment {
        override fun appendSql(sb: StringBuilder) = expr.appendToSql(sb)
      }

      class Raw(private val columnName: String) : Assignment {
        override fun appendSql(sb: StringBuilder) {
          sb.append(columnName)
            .append("=?")
        }
      }
    }

    private val assignments = ArrayList<Assignment>(1)

    internal constructor(
      parent: UpdateSqlNode,
      firstUpdate: Expr
    ) : this(parent) {
      add(firstUpdate)
    }

    internal constructor(
      parent: UpdateSqlNode,
      columnName: String,
      value: String?
    ) : this(parent) {
      addRaw(columnName, value)
    }

    override fun appendSql(sb: StringBuilder) {
      sb.append("SET ")
      assignments.forEachIndexed { index, assignment ->
        if (index > 0) {
          sb.append(',')
        }
        assignment.appendSql(sb)
      }
    }

    private fun add(expr: Expr): Set<T> = apply {
      assignments += Assignment.Expression(expr)
      expr.addArgs(updateBuilder.args)
    }

    private fun addRaw(
      columnName: String,
      value: String?
    ): Set<T> = apply {
      assignments += Assignment.Raw(columnName)
      updateBuilder.args += value
    }

    /**
     * Update a not nullable column with new not nullable value.
     *
     * @param column Column to update. This param must be one of annotation processor
     * generated column objects that corresponds to column in a database table
     * @param value A new value to set for updated column
     * @param V Value type
     * @param R Column return type
     * @param ET Column equivalent type
     * @return SQL UPDATE statement builder
     */
    @CheckResult
    fun <V, R, ET> set(
      column: Column<V, R, ET, T, NotNullable>,
      value: V & Any
    ): Set<T> = add(UpdateColumn(column) IS value)

    /**
     * Update a nullable column with new nullable value.
     *
     * @param column Column to update. This param must be one of annotation processor
     * generated column objects that corresponds to column in a database table
     * @param value A new value to set for updated column
     * @param V Value type
     * @param R Column return type
     * @param ET Column equivalent type
     * @return SQL UPDATE statement builder
     */
    @CheckResult
    fun <V, R, ET> setNullable(
      column: Column<V, R, ET, T, Nullable>,
      value: V?
    ): Set<T> = add(UpdateColumn(column).isNullable(value))

    /**
     * Update a column with the value of another column.
     *
     * @param column Column to update. This param must be one of annotation processor
     * generated column objects that corresponds to column in a database table
     * @param assignmentColumn Column who's value to assign to `column`.
     * This param must be one of annotation processor generated column objects that corresponds to column in a database
     * table
     * @param V Column value type
     * @param R Column return type
     * @param ET Column equivalent type
     * @param N Column nullability
     * @return SQL UPDATE statement builder
     */
    @CheckResult
    fun <V, R, ET, N> set(
      column: Column<V, R, ET, T, N>,
      assignmentColumn: Column<*, *, out ET, *, in N>
    ): Set<T> = add(UpdateColumn(column) IS assignmentColumn)

    /**
     * Update a column with the value of inner SELECT statement result.
     *
     * @param column Column to update. This param must be one of annotation processor
     * generated column objects that corresponds to column in a database table
     * @param select Inner select statement who's result will be assigned to `column`.
     * Select statement must return 1x1 result set.
     * @param V Column value type
     * @param R Column return type
     * @param ET Column equivalent type
     * @param N Column nullability
     * @return SQL UPDATE statement builder
     */
    @CheckResult
    fun <V, R, ET, N> set(
      column: Column<V, R, ET, T, N>,
      select: SelectNode<out ET, Select1, in N>
    ): Set<T> = add(UpdateColumn(column) IS select)

    /**
     * Update a column with new value.
     *
     * @param columnName Column to update
     * @param value A new value to set for updated column
     * @return SQL UPDATE statement builder
     */
    @CheckResult
    fun set(
      columnName: String,
      value: String?
    ): Set<T> = addRaw(columnName, value)

    /**
     * Define SQL UPDATE statement WHERE clause.
     *
     * @param expr WHERE clause expression
     * @return SQL UPDATE statement builder
     */
    @CheckResult
    fun where(expr: Expr) = Where(
      parent = this,
      expr = expr
    )

    /**
     * Define SQL UPDATE statement WHERE clause.
     *
     * @param whereClause WHERE clause
     * @param whereArgs WHERE clause arguments
     * @return SQL UPDATE statement builder
     */
    @CheckResult
    fun where(
      whereClause: String,
      vararg whereArgs: String?
    ) = RawWhere(
      parent = this,
      clause = whereClause,
      args = whereArgs
    )
  }

  /** Builder for SQL UPDATE statement. */
  class Where internal constructor(
    parent: UpdateSqlNode,
    private val expr: Expr
  ) : UpdateNode(parent) {
    init {
      expr.addArgs(updateBuilder.args)
    }

    override fun appendSql(sb: StringBuilder) {
      sb.append("WHERE ")
      expr.appendToSql(sb)
    }
  }

  /** Builder for SQL UPDATE statement. */
  class RawWhere internal constructor(
    parent: UpdateSqlNode,
    private val clause: String,
    args: Array<out String?>
  ) : UpdateNode(parent) {
    init {
      updateBuilder.args.addAll(args)
    }

    override fun appendSql(sb: StringBuilder) {
      sb.append("WHERE ")
        .append(clause)
    }
  }
}
